packages/push/src/evaluation/cases.rs

```diff
@@ -88,35 +236,176 @@ impl<T, Input> WithTargetFn<Input> for T
+    ///
+    /// assert_eq!(cases.len(), 3);
+    /// ```
+    #[must_use]
+    pub fn len(&self) -> usize {
+        self.cases.len()
     }
 }
 
-impl<Input, Output, C> FromIterator<C> for Cases<Input, Output>
-where
-    C: Into<Case<Input, Output>>,
-{
-    fn from_iter<T: IntoIterator<Item = C>>(iter: T) -> Self {
-        Self {
-            cases: iter.into_iter().map(Into::into).collect(),
-        }
+// ------------------
+// ITERATE OVER CASES
+// ------------------
+
+impl<Input, Output> Cases<Input, Output> {
+    /// Creates an iterator over the set of cases.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use push::evaluation::cases::{Case, Cases, WithTargetFn};
+    /// #
+    /// let inputs = ["Hello", "People"];
+    /// let cases = inputs.with_target_fn(|s| s.len());
+    ///
+    /// let mut iter = cases.iter();
+    ///
+    /// assert_eq!(iter.next(), Some(&Case::new("Hello", 5)));
+    /// assert_eq!(iter.next(), Some(&Case::new("People", 6)));
+    /// assert_eq!(iter.next(), None);
+    /// ```
+    pub fn iter(&self) -> impl Iterator<Item = &Case<Input, Output>> {
+        // std::slice::Iter<Case<Input, Output>> {
+        self.cases.iter()
+    }
+
+    /// Creates a iterator over mutable instances of the cases.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use push::evaluation::cases::{Case, Cases};
+    /// #
+    /// let first_case = Case::new("Hello", 5);
+    /// let mut second_case = Case::new("People", 6);
+    /// let mut cases = Cases::new().with_case(first_case).with_case(second_case);
+    ///
+    /// for c in cases.iter_mut() {
+    ///     (*c).output *= 2;
+    /// }
+    ///
+    /// let mut iter = cases.iter();
+    ///
+    /// assert_eq!(iter.next(), Some(&Case::new("Hello", 10)));
+    /// assert_eq!(iter.next(), Some(&Case::new("People", 12)));
+    /// assert_eq!(iter.next(), None);
+    /// ```
+    pub fn iter_mut(&mut self) -> std::slice::IterMut<Case<Input, Output>> {
+        self.cases.iter_mut()
+    }
+
+    /// Creates an iterator over the inputs for this set of cases.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use push::evaluation::cases::Cases;
+    /// #
+    /// let inputs = ["this", "and", "those"];
+    /// let cases = Cases::from_inputs(inputs, |s| s.len());
+    ///
+    /// assert!(cases.inputs().eq(&inputs));
+    /// ```
+    pub fn inputs(&self) -> impl Iterator<Item = &Input> {
+        self.cases.iter().map(|c| &c.input)
+    }
+
+    /// Creates an iterator over the outputs for this set of cases.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use push::evaluation::cases::Cases;
+    /// #
+    /// let inputs = ["this", "and", "those"];
+    /// let cases = Cases::from_inputs(inputs, |s| s.len());
+    ///
+    /// assert!(cases.outputs().eq(&[4, 3, 5]));
+    /// ```
+    pub fn outputs(&self) -> impl Iterator<Item = &Output> {
+        self.cases.iter().map(|c| &c.output)
     }
```

Comment:
if we have `.inputs()` and `.outputs()` shouldn't we have `.inputs_mut()` and `.outputs_mut()` as well as `.into_inputs()` and `.into_outputs()` as well?

Reply:
I like the idea of `.*_mut()`, but do we want/need `.into_*()`? I'm not sure when we'd want to convert a set of `Cases` into their inputs or outputs, losing/consuming the original set of cases. Is this one of those things we should do anyway to be "completist"?